sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala

```
@@ -196,4 +194,107 @@ class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)
+    def exists(path: Path): Boolean
+    def open(path: Path): FSDataInputStream
+    def create(path: Path): FSDataOutputStream
+    def rename(srcPath: Path, destPath: Path): Unit
+    def deleteOnExit(path: Path): Unit
+  }
+
+  /** Implementation of FileManager using newer FileContext API */
+  class FileContextManager(path: Path, hadoopConf: Configuration) extends FileManager {
+    private val fc = if (path.toUri.getScheme == null) {
+      FileContext.getFileContext(hadoopConf)
+    } else {
+      FileContext.getFileContext(path.toUri, hadoopConf)
+    }
+
+    override def list(path: Path, filter: PathFilter): Array[FileStatus] = {
+      fc.util.listStatus(path, filter)
+    }
+
+    override def rename(srcPath: Path, destPath: Path): Unit = {
+      fc.rename(srcPath, destPath)
+    }
+
+    override def mkdirs(path: Path): Unit = {
+      fc.mkdir(path, FsPermission.getDirDefault, true)
+    }
+
+    override def open(path: Path): FSDataInputStream = {
+      fc.open(path)
+    }
+
+    override def create(path: Path): FSDataOutputStream = {
+      fc.create(path, EnumSet.of(CreateFlag.CREATE))
+    }
+
+    override def exists(path: Path): Boolean = {
+      fc.util().exists(path)
+    }
+
+    override def deleteOnExit(path: Path): Unit = {
+      fc.deleteOnExit(path)
+    }
+  }
+
+  /** Implementation of FileManager using older FileSystem API */
+  class FileSystemManager(path: Path, hadoopConf: Configuration) extends FileManager {
+    private val fs = if (path.toUri.getScheme == null) {
+      FileSystem.get(hadoopConf)
+    } else {
+      FileSystem.get(path.toUri, hadoopConf)
+    }
+
+    override def list(path: Path, filter: PathFilter): Array[FileStatus] = {
+      fs.listStatus(path, filter)
+    }
+
+    override def rename(srcPath: Path, destPath: Path): Unit = {
+      if (fs.exists(destPath)) {
+        throw new FileAlreadyExistsException(s"File already exists: $destPath")
+      }
+      fs.rename(srcPath, destPath)
```

Unlike `fc.rename`, `fs.rename` will return `false` if it fails. We should throw an exception for that.